src/core/Map.cpp

```diff
@@ -94,7 +94,23 @@ sc2::Point3D Cluster::Center() const {
 }  // namespace
 
 Expansion::Expansion(const sc2::Point3D& town_hall_location_):
-    town_hall_location(town_hall_location_), owner(Owner::NEUTRAL) {
+    town_hall_location(town_hall_location_), owner(Owner::NEUTRAL),
+    town_hall_tag(sc2::NullTag), worker_tag(sc2::NullTag) {
+}
+
+void Expansion::SetOwner(const sc2::Unit& unit_, Owner owner_) {
```

Comment:
I think it'll be better to use a different approach:
```c++
void Expansion::SetOwner(Owner owner_) {
    owner = owner_;
}

void Expansion::AssignTownHall(const sc2::Unit* unit_) {
  town_hall_tag = unit_.tag;

   if (gAPI->observer().GetCurrentRace() != sc2::Race::Terran)
        worker_tag = sc2::NullTag;
}
```

The idea behind it is to allow mark e.g. enemy expansions without particular town hall tag. This operation is possible with the PR's code, too, but requires e.g. nullptr in sc2::Unit which is not expected. 

What do you think about supporting marking enemy expansions without particular tag?

Reply:
I like having Enemy town_hall_tag so it is easier to check what their health is and if they are upgraded when attempting to attack/harrass.

If player hasn't seen an EnemyTownhall yet, they can mark the region as CONTESTED instead. For enemy starting_location, a worker scout should see the enemy's main townhall fairly easily, but if ramp is walled, other scouting attempts should be continued until successful.

I don't see a reason to mark a region as ENEMY without seeing a townhall, which CONTESTED should do otherwise, and enemy_starting_location is already a known special enemy region without the need for an explicit OWNER mark.